### PR TITLE
Add Fetch and Remove dom in tsconfig

### PR DESCRIPTION
**Pull request by huhuanming**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://facebook.github.io/react-native/docs/network.html>>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by alloy** [comment]
The fetch part looks good to me, but I don’t see any changes related to DOM and tsconfig, what gives?

**Comment by huhuanming** [comment]
I removed dom just now, but all tests of  projects related `react-native`  is failed.

😳 
that would be expedient.

File: types/react-native/globals.d.ts [hunk 1]
```diff
@@ -18,3 +18,5 @@ declare function setImmediate(handler: (...args: any[]) => void): number;
 
 declare function cancelAnimationFrame(handle: number): void;
 declare function requestAnimationFrame(callback: (time: number) => void): number;
+
+declare function fetch(input: RequestInfo, init?: RequestInit): Promise<Response>;
```
Comment (timwangdev) [inline types/react-native/globals.d.ts]: `RequestInfo`, `RequestInit`, `Response` are dom defination and not defined in react-native.

Comment (huhuanming) [reply types/react-native/globals.d.ts]: it is indeed.

It seems that I should remove dom from tsconfig in all  projects related `react-native`

Comment (huhuanming) [reply types/react-native/globals.d.ts]: And add `Websocket`.

**Comment by alloy** [comment]
Hmm. Alas I know very little about the DOM related typings. Paging @newyankeecodeshop, as he recently also did work surrounding DOM typings.

**Comment by newyankeecodeshop** [comment]
I believe this PR won't fully work because it's missing the declarations for all the `fetch` interfaces that are provided in the DOM library. There are two important pieces we need to provide to enable a RN project to use just `ES2015-7`:
1. Provide the complete `fetch` API definitions. (I'm not sure we care about `XmlHttpRequest`?)
2. Provide empty HTML interfaces that the `React` type requires. Since `React` has so many specific behaviors related to HTML element component, the `index.d.ts` in `@types/react` requires the DOM library. However, we can provide empty placeholders to use when a project does not include the DOM.

I have created a gist that includes the missing declarations. You can reference these files in a `global.d.ts` in your project, and it should compile cleanly.
https://gist.github.com/newyankeecodeshop/57141e58880484ca8dc2884d7fa69d0f

I was planning to add these to this project soon, once I had made sure that there are no compatibility issues with existing projects that already include the DOM library.
Thanks for working on this @huhuanming and @alloy

**Review by timwangdev — CHANGES_REQUESTED** [review]
Add this to prevent auto-merge

**Comment by alloy** [comment]
> 1. Provide the complete `fetch` API definitions. (I'm not sure we care about `XmlHttpRequest`?)

Ideally we’d have typings for both. The `fetch` API is afaik a polyfill on top of the native `XmlHttpRequest` implementation, so users may want to prefer that.

**Comment by newyankeecodeshop** [comment]
My rationale for only including `fetch` types was that (1) it's the more "modern" API for networking and (2) if folks are trying to remove DOM dependencies, they're more likely to be using `fetch` than the old fashioned XHR.

I'm not sure what can be done with the `react` typings in regards to separating the HTML-specific parts. I tried unsuccessfully to move the DOM-related parts from `react` to `react-dom`, but that broke lots of the tests. We probably need help from the TypeScript team to introduce a directive that can exclude declarations when a core library is not available.

@alloy Do you think it would be worth logging an issue with `react` typings to see if they can include the placeholder DOM definitions in a similar `globals.d.ts` file?

**Comment by huhuanming** [comment]
I think so.

**Comment by newyankeecodeshop** [comment]
Good news! PR #20005 has been merged and deployed as `@types/react@16.0.8`.
I don't know how to specify dependencies for specific versions of types, but it probably will allow us to change the `tsconfig.json` for this project and remove `dom` from the library list.

**Comment by huhuanming** [comment]
Cool!

I will check it in next day.

**Comment by alloy** [comment]
@newyankeecodeshop Nice job! :clap: I won’t have much time to test this out over the next few days, so for now I’ll defer to @huhuanming to test it out.

**Comment by huhuanming** [comment]
@alloy

**Review by timwangdev — APPROVED** [review]
LGTM

**Comment by alloy** [comment]
Awesome! Great job, y’all! 👏

**Comment by stephenh** [comment]
I'm futzing with the react-native typings, and running just `tsc` in a clean checkout results in:

```
$ tsc
globals.d.ts(24,31): error TS2304: Cannot find name 'RequestInfo'.
globals.d.ts(24,51): error TS2304: Cannot find name 'RequestInit'.
globals.d.ts(24,73): error TS2304: Cannot find name 'Response'.
```

Should this PR have included definitions for them in global.d.ts, given that they're not coming from the now-removed dom library anymore?

**Comment by mhegazy** [comment]
Please add a new PR to include the missing types, and we will review it.

**Comment by newyankeecodeshop** [comment]
I looked into doing this a while back but the type definitions have subtle changes between TypeScript 2.3 - 2.6. I think they changed slightly in 2.8 as well. The problem is that we need to include definitions that are compatible with folks that do import the "dom" library. Otherwise TS will complain about incompatible definitions.
That being said, now that these types require TS 2.6+, perhaps we can come up with compatible definitions. Also, since React Native uses the `whatwg-fetch` polyfill, not all features in those interfaces are necessary to include.

**Comment by stephenh** [comment]
@newyankeecodeshop I'm definitely willing to submit a PR, but it will likely be a naive copy/paste of whatever is in the current TS dom definitions. It sounds like you have a lot more historical context to the potential nuance here; would you like to submit a PR instead? Either is fine with me.

**Comment by newyankeecodeshop** [comment]
I don't have time to get to it this week @stephenh so if you want to give it a shot, go for it. It might be good to add the most minimal interfaces you can to make it compile for you. TypeScript will merge the definitions from the DOM library if people have projects that include it.

**Comment by stephenh** [comment]
I played around with this, but resurrecting the @types/whatwg-fetch definitions before they were deleted/inlined into lib dom (I figured this would most closely match what RN has given they use whatwg-fetch directly).

However, in my project that pulls in the new global.d.ts, I'm getting a lot of duplicate identifier errors between lib.dom and the new global.d.ts definitions. If I remove lib dom (which I should probably, given it's not actually available in RN), then they all go away.

But, AFAICT, it's a requirement to still support (with no errors) projects that do (mistakenly?) pull in lib dom?

I did fix a few duplicate errors by the hack of changing "class Foo" to "interface Foo + declare var Foo". Seems kind of ugly.

However, I can't get rid of duplicate errors for type defs, e.g.:

```typescript
declare type RequestType = "" | "audio" | "font" | "image" | "script" | "style" | "track" | "video";
```

If there a kosher way, similar to the interface+var approach for classes, to work around the duplicate identifier errors for type defs?

I thought just inlining them would work, but for the ~3-4 I've tried, I still get errors because AFAICT they've drifted slightly from what is in lib dom.

This is harder than I thought it would be. :-/ Suggestions?

**Comment by newyankeecodeshop** [comment]
@stephenh I created a PR with the missing interfaces. They are not a complete copy/paste from the standard library since the RN fetch polyfill, `whatwg-fetch`, does not implement the complete specification. Can you try out my types and see if they work in your application?

**Comment by imod** [comment]
@newyankeecodeshop not having `XmlHttpRequest` included in the type definitions anymore, causes quite some issues with e.g. `rxjs/ajax`.

Having a service like this:
```
import { Observable } from "rxjs";
import { ajax, AjaxResponse } from "rxjs/ajax";
import { map } from "rxjs/operators";

export function login(
  username: string,
  password: string
): Observable<string> {
  return ajax.post(
    `http://localhost:8080/login`,
    {
      userId: username,
      password: password
    },
    { "Content-Type": "application/json" }
  ).pipe(map<AjaxResponse,string>(response => (response.response.token)));
}
```

WIll cause errors like this:

```
platypus:my-project domi$ tsc --noEmit
node_modules/rxjs/internal/observable/dom/AjaxObservable.d.ts:16:23 - error TS2304: Cannot find name 'XMLHttpRequest'.

16     createXHR?: () => XMLHttpRequest;
                         ~~~~~~~~~~~~~~
```
here is a very simple example project:  https://github.com/imod/typescript-issue

**Comment by newyankeecodeshop** [comment]
@imod I haven't had a chance to come back to this. I think we can just copy the `XMLHttpRequest` declarations from the TypeScript library into the globals. I will try to get to it, but if you need it now, it's something you can do locally in your project. Create a `global.d.ts` and copy the definitions there.

**Comment by imod** [comment]
@newyankeecodeshop thanks - that's what I did for now, but I wanted to give some feedback so this might get improved for the future

**Comment by newyankeecodeshop** [comment]
@imod I have a branch you can try:
https://github.com/newyankeecodeshop/DefinitelyTyped/tree/rn-xmlhttprequest
If it works I can do a PR from it. Our project doesn't use `rxjs` so I can't test it to make sure it fixes your build. I basically just copied the declarations from the TypeScript 2.9 lib and commented out the stuff that RN doesn't implement.

**Comment by imod** [comment]
@newyankeecodeshop ok, just replaced the content of `node_modules/@types/react-native/globals.d.ts` with the one from your PR and it does work nice now.

**Comment by newyankeecodeshop** [comment]
@imod Here is the PR: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/30718

**Comment by imod** [comment]
@newyankeecodeshop many thanks!!

**Pull request merged**